overlord/devicestate/devicestate_remodel_test.go

```diff
@@ -7294,3 +7294,260 @@ func snapSetupFromTaskSet(ts *state.TaskSet) *snapstate.SnapSetup {
+		return ts, nil
+	})
+	defer restore()
+
+	// set a model assertion
+	s.makeModelAssertionInState(c, "canonical", "pc-model", map[string]interface{}{
+		"architecture": "amd64",
+		"base":         "core20",
+		"grade":        "dangerous",
+		"classic":      "true",
+		"distribution": "ubuntu",
+		"snaps": []interface{}{
+			map[string]interface{}{
+				"name":            "pc-kernel",
+				"id":              snaptest.AssertedSnapID("pc-kernel"),
+				"type":            "kernel",
+				"default-channel": "20",
+			},
+			map[string]interface{}{
+				"name":            "pc",
+				"id":              snaptest.AssertedSnapID("pc"),
+				"type":            "gadget",
+				"default-channel": "20",
+			},
+		},
+	})
+	s.makeSerialAssertionInState(c, "canonical", "pc-model", "serial")
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand:  "canonical",
+		Model:  "pc-model",
+		Serial: "serial",
+	})
+
+	var gadgetYaml = `
+volumes:
+  pc:
+    bootloader: grub
+    structure:
+      - name: ubuntu-seed
+        role: system-seed
+        type: EF,C12A7328-F81F-11D2-BA4B-00A0C93EC93B
+        size: 1G
+      - name: ubuntu-data
+        role: system-data
+        type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+        size: 2G
+`
+
+	gadgetFiles := [][]string{
+		{"meta/gadget.yaml", gadgetYaml},
+	}
+
+	snapstatetest.InstallEssentialSnaps(c, s.state, "core20", gadgetFiles, nil)
+
+	newModel := s.brands.Model("canonical", "pc-model", map[string]interface{}{
+		"architecture": "amd64",
+		"base":         "core20",
+		"grade":        "dangerous",
+		"revision":     "1",
+		"classic":      "true",
+		"distribution": "ubuntu",
+		"snaps": []interface{}{
+			map[string]interface{}{
+				"name":            "pc-kernel",
+				"id":              snaptest.AssertedSnapID("pc-kernel"),
+				"type":            "kernel",
+				"default-channel": "21/edge",
+			},
+			map[string]interface{}{
+				"name":            "pc",
+				"id":              snaptest.AssertedSnapID("pc"),
+				"type":            "gadget",
+				"default-channel": "21/stable",
+			},
+			map[string]interface{}{
+				"name":            "core20",
+				"id":              snaptest.AssertedSnapID("core20"),
+				"type":            "base",
+				"default-channel": "latest/edge",
+			},
+		},
+	})
+
+	chg, err := devicestate.Remodel(s.state, newModel, nil, nil, devicestate.RemodelOptions{})
+	c.Assert(err, IsNil)
+
+	c.Check(chg.Summary(), Equals, "Refresh model assertion from revision 0 to 1")
+
+	tasks := chg.Tasks()
+
+	// 3 snaps (3 tasks for each) + recovery system (2 taskss) + set-model
+	c.Check(tasks, HasLen, 3*3+2+1)
+
+	taskKinds := make([]string, 0, len(tasks))
+	for _, t := range tasks {
+		taskKinds = append(taskKinds, t.Kind())
+	}
+
+	c.Check(taskKinds, DeepEquals, []string{
+		"fake-download", "validate-snap", "fake-update",
+		"fake-download", "validate-snap", "fake-update",
+		"fake-download", "validate-snap", "fake-update",
+		"create-recovery-system", "finalize-recovery-system",
```

Comment:
But we don't support creating a recovery system for hybrid yet, or am I getting that wrong? I wonder because if that is the case maybe the result of this test should be an error so we do not try, until we really support this. cc @pedronis

Reply:
do we have a TODO for the spread test for that scenario? we need to keep in mind to add that one

I don't have a strong opinion about the unit tests, either they pass already, or we add a TODO? the questions there if whether we fear they are realistic because the real support will diverge

Reply:
Yes, we have a TODO in the spread test. And as for the unit test, I think it should stay pretty similar. This test in particular only checks that we create the correct set of tasks.